### PR TITLE
Add settings option for double speed CPU

### PR DESCRIPTION
**Pull request by pau-tomas**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our [guidelines](/chrismaltby/gb-studio/blob/develop/.github/COMMIT_MESSAGE_GUIDELINES.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Add settings option for double speed CPU. 

* **What is the current behavior?** (You can also link to an open issue here)

See https://github.com/chrismaltby/gb-studio/issues/289 for more details

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

File: src/lib/compiler/makeBuild.js [hunk 1]
```diff
@@ -106,6 +106,9 @@ const makeBuild = ({
     } else {
       gameHeader = gameHeader.replace(/#define CUSTOM_COLORS/g, '');
     }
+    if(!settings.gbcFastCPUEnabled) {
+      gameHeader = gameHeader.replace(/#define FAST_CPU/g, '');
+    }
     await fs.writeFile(`${buildRoot}/include/game.h`, gameHeader, "utf8");
 
     // Remove GBC Rombyte Offset from Makefile (OSX/Linux) if custom colors not enabled
```
Comment (pau-tomas) [inline src/lib/compiler/makeBuild.js]: 🤔 I think this need to be changed to remove it when both custom colors and fast cpu enabled are not enabled.

**Comment by chrismaltby** [comment]
Seems to work great! Is this ready to merge?

With this enabled I'm not noticing the music skips while switching scenes anymore. Are there any downsides to using the feature at all?

If there's no reason to ever turn it off, I might rethink some of the UI around GBC support at some point and just make a single setting to turn on/off GBC support that states 2x CPU speed as a benefit.

Thanks @pau-tomas and @RichardULZ

**Comment by RichardULZ** [comment]
Reasons not to use it, 
Slight hit to battery power on real hardware, may be harder to emulate on slow devices like phones, not much else, may be best to leave it on.

Reasons to still make a GB only game.
Swappable color palettes when running a gb only game on a real CGB, and faster to emulate on slow devices.

Custom palettes must be enabled when using any CGB feature, so bundling 2x cpu would work great.

I have some modifications to the load scene function that further improve skipping audio in fast cpu mode, such as turning on the display after graphics are loaded, as wait vblank is immediate while screen off, and can cause a complex song to jump 3-4 notes ahead.

With double cpu, Small scenes have no hitch, large scenes have a tiny hitch, text boxes produce no hitch.

Could call GBT twice for large scenes as a crude fix.

**Comment by pau-tomas** [comment]
@RichardULZ do you think this is ready to merge? I can take it out from draft if that's the case.

**Comment by RichardULZ** [comment]
@pau-tomas 
There's a tiny issue with the settings page when gb studio is in a **small window**, 
Enable, palette, then enable Double CPU, This can create an empty space covering the bottom of the screen, pushing the title bar off screen. Only shows when gb studio is fairly small. likely due to reusing some code from opening the palette picker. Only tested on windows 10.

The palette picker should probably be open whenever double CPU speed is on though, may cause confusion if they set a palette, disabled it, but find the palette is still being used.

Other option is combining double CPU speed with GBC palette picker, 
The only real downside may be on slower phones, but that's only noticeable when music is playing, and anyone with music in game will likely turn this setting on...

So, 2 options.
- Fix the current settings page when small, maybe force colors to stay open too... 
- Rename `Export With Custom Palette` to also say `& Double CPU Speed (GBC)` as @chrismaltby suggested, since there is very little downside.

**Comment by pau-tomas** [comment]
@chrismaltby I noticed that Game Boy isn't mentioned anywhere in the app so I was wondering if it's due to ©️  concerns or if it would be ok to rename the `Colors` settings section to `Game Boy Color options` and add the checkboxes for Custom Palette and Double CPU Speed? Maybe having a not stating that enabling any of the options will generate roms only compatible with GBC.

**Comment by RichardULZ** [comment]
@pau-tomas I believe the games that use GBC features are still compatible with original Gameboy hardware, as the system type is checked before using those functions, The extra note may not be necessary.

**Comment by pau-tomas** [comment]
Ah, my bad. Scratch the extra note then.

I guess we won't get into "GBC only" territory until GBS provides a way to use the extra tile maps  and so.

**Comment by RichardULZ** [comment]
Have done some testing, there are practically no downsides to having Double Speed CPU combined with GBC palette.

Testing on a low end phone that stutters with audio, performance was near identical between 1x and 2x speed roms.
As for battery life on real hardware, https://azug.minpet.unibas.ch//~lukas/GBprojects/gbFAQ.html#PowerUsage 
You'll use an extra 7-8 milliamps of power, 10%, Not worth worrying about if it means better sound.

**Comment by chrismaltby** [comment]
Coming back to this I think it might be better to let people decide if they want double speed mode enabled and keep it as separate option.

Just made a few updates to reorganise these settings and fix some of the issues that @RichardULZ was seeing.

- The display issue where the toolbar disappeared when toggling the checkbox on a small screen was one of those weird CSS bugs that fixed itself by adding `transform: translate3d(0,0,0)` to the settings page container forcing hardware acceleration for that `div`

- The double speed mode now uses a default GBC palette if you enable it without checking "custom palette"

- Updated the default GBC palette to match the suggestion from @proximitysound in #308

I think this might be ready to merge now, @pau-tomas do you want to take the pull request out of draft?

**Pull request merged**